solutions/usaco-743.mdx

```diff
@@ -0,0 +1,85 @@
+---
+id: usaco-743
+source: USACO Silver 2017 Open
+title: Modern Art 2
+author: Sofia Yang
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_art2_gold_open17.html)
+
+
+### Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class ModernArt2 {
+
+    public static void main(String[] args) throws IOException {
+        BufferedReader in = new BufferedReader(new FileReader("art2.in"));
+        PrintWriter out = new PrintWriter(new FileWriter("art2.out"));
+        int N = Integer.parseInt(in.readLine());
+
+        int[] start = new int[N];
+        int[] end = new int[N];
+        Arrays.fill(start, Integer.MAX_VALUE);
+        Arrays.fill(end, Integer.MAX_VALUE);
+
+        int[] color = new int[N];
+
+        for (int i = 0; i < N; i++) {
+            color[i] = Integer.parseInt(in.readLine());
+        }
+
+        for (int i = 0; i < N; i++) {
+            // Update the first and last appearance of the color.
+            start[color[i]] = Integer.min(start[color[i]], i);
+            end[color[i]] = Integer.max(end[color[i]], i);
+        }
+
+        int ans = 0;
+        // The contents of the stack are the colors of the currently open segments.
+        Stack<Integer> stack = new Stack<>();
+        for (int idx = 0; idx < N; idx++) {
+            if (color[idx] ==0) {
+                continue;
+            }
+            int curColor = color[idx];
+
+            // Process the start point of the color segment.
+            if (idx == start[curColor]) {
+                stack.push(curColor);
+                ans = Integer.max(ans, stack.size());
+            }
+
+            /*
+             * Invalid because there are 2 partially intersecting segments.
+             * For instance if we have a finished painting of 1 1 2 1 2,
+             * it is impossible if we can only paint 1 once.
+             */
+            if (stack.get(stack.size() - 1) != curColor) {
+                out.println(-1);
+                out.close();
+                break;
+            }
+
+            // Remove the color from the stack since the segment has closed.
+            if (idx == end[curColor]) {
+                stack.pop();
+            }
+        }
+
+        // Subtract the color '0' segment.
+        out.println(ans - 1);
```

Comment:
If i am interpreting your code correctly I don't think the '0' segment exists anymore

Reply:
sorry, i'll delete the comment